OBJC_INTEROP.md

```diff
@@ -0,0 +1,124 @@
+| Method | Method | Method | [note](#top-level-functions-and-properties) |
+| `@Throws` | `throws` | `error:(NSError**)error` | [note](#errors-and-exceptions) |
+| Extension | Extension | Category member | [note](#category-members) |
+| Companion object member <- | Class method or property | Class method or property |  |
+| `Singleton` | `Singleton()`  | `[Singleton singleton]` | |
+| Primitive type | Primitive type / `NSNumber` | | [note](#number) |
+| `String` | `String` | `NSString` | |
+| `String` | `NSMutableString` | `NSMutableString` | [note](#nsmutablestring) |
+| `List` | `Array` | `NSArray` | |
+| `MutableList` | `NSMutableArray` | `NSMutableArray` | |
+| `Set` | `Set` | `NSSet` | |
+| `MutableSet` | `NSMutableSet` | `NSMutableSet` | [note](#mutable-collections) |
+| `Map` | `Dictionary` | `NSDictionary` | |
+| `MutableMap` | `NSMutableDictionary` | `NSMutableDictionary` | [note](#mutable-collections) |
+| Function type | Function type | Block pointer type | |
+
+### Name translation
+
+Objective-C classes are imported into Kotlin with their original names.
+Protocols are imported as interfaces with `Protocol` name suffix,
+i.e. `@protocol Foo` -> `interface FooProtocol`.
+These classes and interfaces are placed into package [specified in build configuration](#usage)
+(`platform.*` packages for preconfigured system frameworks).
+
+Names of Kotlin classes and interfaces are prefixed when imported to Swift/Objective-C.
+The prefix is derived from the framework name.
+
+### Top-level functions and properties
+
+Top-level Kotlin functions and properties are accessible as members of a special class.
+Each Kotlin package is translated into such a class.
+
+### Errors and exceptions
+
+Kotlin has no concept of checked exceptions, all Kotlin exceptions are unchecked.
+Swift has only checked errors. So if Swift or Objective-C code calls Kotlin method
+which throws an exception to be handled, then Kotlin method should be marked
+with `@Throws` annotation. In this case all Kotlin exceptions
+(except for instances of `Error`, `RuntimeException` and subclasses) are translated to
+Swift error/`NSError`.
+
+Note that the opposite translation is not implemented yet:
+Swift/Objective-C error-throwing methods aren't imported to Kotlin as
+exception-throwing.
+
+### Category members
+
+Members of Objective-C categories and Swift extensions are imported to Kotlin
+as extensions. That's why these declarations can't be overridden in Kotlin.
+And extension initializers aren't available as Kotlin constructors.
+
+### Number
+
+`NSNumber` is not automatically translated to Kotlin primitive types
+when used as Swift/Objective-C parameter type or return value.
+
+### NSMutableString
+
+`NSMutableString` Objective-C class is not available from Kotlin.
+All instances of `NSMutableString` are copied when passed to Kotlin.
+
+### Mutable collections
+
+Every Kotlin `MutableSet` is `NSMutableSet`, however the opposite is not true.
```

Comment:
Please mention explicitly that all other collections, such as Map and List are properly mapped.

Reply:
Actually while writing these paragraphs I was considering them to be read just as notes to the table, after following the corresponding links.
Now I see that they look very silly when being read as self-contained descriptions. So I suppose some of the paragraphs should be expanded.